### ISSUE TITLE
False positive for F841

### ISSUE THREAD

**Issue by bryevdv** [charliermarsh/ruff#820]
Today the latest version of ruff (0.0.128) started  complaining about this test code:
```python
    def test_func_default_with_counter(self) -> None:
        counter = 0
        def next_value() -> int:
            nonlocal counter
            counter += 1
            return counter
        class HasFuncDefaultInt(Model):
            value = Int(default=next_value)
        obj1 = HasFuncDefaultInt()
        obj2 = HasFuncDefaultInt()
        assert obj1.value + 1 == obj2.value
```
with 
> tests/unit/bokeh/test_objects.py:291:9: F841 Local variable `counter` is assigned to but never used

It seems  the lasts`ruff` does not understand the usage with `nonlocal`. Previous versions of `ruff` (and `flake8`) do not warn on this code. 

(this isn't code we'd ever use in the library itself, but we do have to handle whatever crazy cases users might throw at us, and this is a test for one of those)

**Comment by charliermarsh** [comment]
I’ll take a look at this tonight! Thanks for reporting.

**Comment by bryevdv** [comment]
I just realized I didn't actually provide a proper MRE:
```python
def foo():
    counter = 0
    def next_value() -> int:
        nonlocal counter
        counter += 1
        return counter
    a = next_value()
    b = next_value()
    return a, b

print(f"(a, b): {foo()!r}")
```
with this setup:

https://github.com/bokeh/bokeh/blob/46b77a3a8d872237f333b7c7193887016d9ce507/pyproject.toml#L111-L124

**Comment by charliermarsh** [comment]
Thanks! As I expected this was introduced in https://github.com/charliermarsh/ruff/pull/800.

**Comment by charliermarsh** [comment]
The [Chesterton's Fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence) was strong in this one:

<img width="812" alt="Screen Shot 2022-11-19 at 7 01 50 PM" src="https://user-images.githubusercontent.com/1309177/202876328-a0b7d64b-3a16-4791-a009-5bcb76a0d175.png">

**Comment by charliermarsh** [comment]
This is going out in [v0.0.129](https://github.com/charliermarsh/ruff/releases/tag/v0.0.129) (building now).
